(PUP-7571) Relax version ranges used by puppet module tool

thallgren opened the pull request:
This commit relaxes all version ranges used by the puppet module tool
so that they include versions with pre-releases. This mimics the broken
behavior of SemanticVersion 0.1.4.

This commit is not intended to be the final solution. Its intended as
a "does this solve the PEZ problem" test.

puppetcla commented:
CLA signed by all contributors.

joshcooper commented:
This doesn't solve the PEZ issue:

```
[root@etytst2sbinkrvq puppet]# git rev-parse HEAD
30773ba2ba464dd3edbc64a8459b3caa731edc79
[root@etytst2sbinkrvq puppet]# /opt/puppetlabs/puppet/bin/bundle exec puppet module list
Warning: Setting app_management is deprecated.
   (at /root/puppet/lib/puppet/settings.rb:1159:in `issue_deprecation_warning')
Warning: Module 'puppetlabs-pe_infrastructure' (v2017.2.0-rc2) fails to meet some dependencies:
  'puppetlabs-pe_install' (v2017.3.0-rc0-16-g9d26022) requires 'puppetlabs-pe_infrastructure' (>= 2017.2.0)
  'puppetlabs-pe_manager' (v2017.2.0-rc1-6-g849d56e) requires 'puppetlabs-pe_infrastructure' (>= 2017.2.0)
  'puppetlabs-puppet_enterprise' (v2017.3.0-rc0-80-g3063b46) requires 'puppetlabs-pe_infrastructure' (>=2017.2.0)
Warning: Module 'puppetlabs-pe_install' (v2017.3.0-rc0-16-g9d26022) fails to meet some dependencies:
  'puppetlabs-pe_manager' (v2017.2.0-rc1-6-g849d56e) requires 'puppetlabs-pe_install' (>= 2017.2.0)
Warning: Module 'puppetlabs-pe_puppet_authorization' (v2016.2.0-rc1) fails to meet some dependencies:
  'puppetlabs-puppet_enterprise' (v2017.3.0-rc0-80-g3063b46) requires 'puppetlabs-pe_puppet_authorization' (>=2016.2.0)
Warning: Module 'puppetlabs-puppet_enterprise' (v2017.3.0-rc0-80-g3063b46) fails to meet some dependencies:
  'puppetlabs-pe_accounts' (v2017.1.0-rc0) requires 'puppetlabs-puppet_enterprise' (>= 3.4.0)
  'puppetlabs-pe_concat' (v1.1.2) requires 'puppetlabs-puppet_enterprise' (>= 3.4.0)
  'puppetlabs-pe_install' (v2017.3.0-rc0-16-g9d26022) requires 'puppetlabs-puppet_enterprise' (>= 2017.2.0)
  'puppetlabs-pe_postgresql' (v2016.5.0) requires 'puppetlabs-puppet_enterprise' (>=2016.5.0)
  'puppetlabs-pe_puppet_authorization' (v2016.2.0-rc1) requires 'puppetlabs-puppet_enterprise' (>= 2015.3.0)
  'puppetlabs-pe_r10k' (v2016.2.0) requires 'puppetlabs-puppet_enterprise' (> 3.8.0)
  'puppetlabs-pe_repo' (v2017.3.0-rc0-12-g58b800e) requires 'puppetlabs-puppet_enterprise' (>= 2017.2.0)
/etc/puppetlabs/code/environments/production/modules (no modules installed)
/etc/puppetlabs/code/modules (no modules installed)
/opt/puppetlabs/puppet/modules
├── puppetlabs-pe_accounts (v2017.1.0-rc0)
├── puppetlabs-pe_concat (v1.1.2)
├── puppetlabs-pe_hocon (v2016.2.0)
├── puppetlabs-pe_infrastructure (v2017.2.0-rc2)  invalid
├── puppetlabs-pe_inifile (v2016.5.0)
├── puppetlabs-pe_install (v2017.3.0-rc0-16-g9d26022)  invalid
├── puppetlabs-pe_java_ks (v2016.4.0)
├── puppetlabs-pe_manager (v2017.2.0-rc1-6-g849d56e)
├── puppetlabs-pe_nginx (v2017.1.0)
├── puppetlabs-pe_postgresql (v2016.5.0)
├── puppetlabs-pe_puppet_authorization (v2016.2.0-rc1)  invalid
├── puppetlabs-pe_r10k (v2016.2.0)
├── puppetlabs-pe_razor (v1.0.0-5-gb37afe3)
├── puppetlabs-pe_repo (v2017.3.0-rc0-12-g58b800e)
├── puppetlabs-pe_staging (v2015.3.0)
├── puppetlabs-pe_support_script (v2016.5.0-51-g67d4059)
└── puppetlabs-puppet_enterprise (v2017.3.0-rc0-80-g3063b46)  invalid
```

As I mentioned in the ticket, it's a bigger issue than the prerelease version string. The number of commits after the git tag and git describe in the version string, e.g. `-80-g3063b46` are causing issues.

thallgren commented:
OK. But then there are two separate issues.

1. ">= 2017.2.0" is expected to include version "2017.2.0-rc2. That makes no sense at all and must be fixed in PEZ. A pre-release always comes before the release (hence the name). The constraint must be lowered to something that includes the pre-release. ">2017.1.0" or ">=2017.2.0.0" would both work.

2. The problem of not being able to appoint pre-releases at all, is fixed. The range ">= 2017.2.0" does include "2017.2.1-rc2" or "2017.3.0-rc2".

thallgren commented:
Ah, and then I saw it. "2017.3.0-rc0-80-g3063b46" is not considered a valid version because it contains multiple dashes. Well, that's easy to fix.

thallgren commented:
@joshcooper, I missed a couple of places where VersionRange was used (most notably the module list command). I've updated the PR now and tested it with `puppet module install` on the machine that you had set up for me. It no longer shows any versions as 'invalid'.

joshcooper commented:
Transient appveyor failure, retriggering.

joshcooper commented on lib/puppet/module.rb:
```diff
@@ -58,7 +58,7 @@ def self.is_module_namespaced_name?(name)
   attr_accessor :source, :author, :version, :license, :summary, :description, :project_page
 
-  def initialize(name, path, environment)
+  def initialize(name, path, environment, include_prerelease = true)
```
Should this default be `false`?

thallgren replied:
Ideally, it should be controlled by a setting or option somewhere. The reason it is true here is that it resolves the PEZ problem and ensures backward compatibility for modules that might rely on the old behavior. So, pending a decision on how the more strict ranges should be handled from a migration perspective, I think it needs to be true for now.

joshcooper replied:
Ah, I was asking because everywhere else we default the optional parameter to false:

https://github.com/puppetlabs/puppet/pull/5883/files#diff-40bfd58bbeef98e0cdbe69d6a84d17e5R120
https://github.com/puppetlabs/puppet/pull/5883/files#diff-ef9407d2777b719bcec044e1e15ef5eaR295
https://github.com/puppetlabs/puppet/pull/5883/files#diff-f0460bb812c9b3d7dcf6d83c6eb70310R57
https://github.com/puppetlabs/puppet/pull/5883/files#diff-d2a368760925bf50f080e11ac668aef2R46

And then we set it to true in the various applications that call into the PMT related code, e.g. https://github.com/puppetlabs/puppet/pull/5883/files#diff-a410247f3ed759df4a60550e388bbd6cR76.

thallgren replied:
That's correct. The intention is to have everything semver compliant unless told differently from PMT.

joshcooper replied:
> That's correct. The intention is to have everything semver compliant

That's why I was asking about the `Puppet::Module` initializer:

```ruby
def initialize(name, path, environment, include_prerelease = true)
```

Shouldn't that be `false`?

thallgren replied:
No, it shouldn't. That's one of the places that I identified as a "root", i.e. where we need to make a decision.

joshcooper commented:
Could you add tests to exercise the issues found in PEZ?

thallgren commented:
I rewrite the implementation and changed the following:
1. The changes to vendored semantic_puppet were moved into a separate commit.
2. Changed the `include_prerelease` flag to the shorter and more generic `strict_semver` which defaults to `true` in the implementation but to `false` for `puppet module`
3. Added a proper `--strict-semver` flag to the relevant `puppet module` sub-commands.
4. Added tests to assert that the flag (and the default) performs as expected.

Magisus commented:
Verified that this appears to work on a machine set up for PE testing:
```
[root@etytst2sbinkrvq puppet]# /opt/puppetlabs/puppet/bin/bundle exec puppet module list
Warning: Setting app_management is deprecated.
   (at /root/puppet/lib/puppet/settings.rb:1159:in `issue_deprecation_warning')
/etc/puppetlabs/code/environments/production/modules (no modules installed)
/etc/puppetlabs/code/modules (no modules installed)
/opt/puppetlabs/puppet/modules
├── puppetlabs-pe_accounts (v2017.1.0-rc0)
├── puppetlabs-pe_concat (v1.1.2)
├── puppetlabs-pe_hocon (v2016.2.0)
├── puppetlabs-pe_infrastructure (v2017.2.0-rc2)
├── puppetlabs-pe_inifile (v2016.5.0)
├── puppetlabs-pe_install (v2017.3.0-rc0-16-g9d26022)
├── puppetlabs-pe_java_ks (v2016.4.0)
├── puppetlabs-pe_manager (v2017.2.0-rc1-6-g849d56e)
├── puppetlabs-pe_nginx (v2017.1.0)
├── puppetlabs-pe_postgresql (v2016.5.0)
├── puppetlabs-pe_puppet_authorization (v2016.2.0-rc1)
├── puppetlabs-pe_r10k (v2016.2.0)
├── puppetlabs-pe_razor (v1.0.0-5-gb37afe3)
├── puppetlabs-pe_repo (v2017.3.0-rc0-12-g58b800e)
├── puppetlabs-pe_staging (v2015.3.0)
├── puppetlabs-pe_support_script (v2016.5.0-51-g67d4059)
└── puppetlabs-puppet_enterprise (v2017.3.0-rc0-80-g3063b46)
```

Magisus commented on lib/puppet/vendor/semantic_puppet/lib/semantic_puppet/version_range.rb:
```diff
@@ -219,10 +220,13 @@ def self.parse_lteq_version(expr)
     #   Creates a new instance based on parsed content. For internal use only
     #   @param ranges [Array<AbstractRange>] the ranges to include in this range
     #   @param string [String] the original string representation that was parsed to produce the ranges
+    #   @param include_prerelease [Boolean] `true` if prereleases should be included even when not explicitly appointed
```
This comment seems to be out of date.

Magisus commented on lib/puppet/vendor/semantic_puppet/lib/semantic_puppet/version_range.rb:
```diff
@@ -1,422 +1,758 @@
+    #   Creates a new instance based on parsed content. For internal use only
+    #   @param ranges [Array<AbstractRange>] the ranges to include in this range
+    #   @param string [String] the original string representation that was parsed to produce the ranges
+    #   @param include_prerelease [Boolean] `true` if prereleases should be included even when not explicitly appointed
```
This comment appears to be out of date.

thallgren replied:
Oops. Forgot that one. Added a commit with a fix.

joshcooper commented on lib/puppet/face/module/install.rb:
```diff
@@ -119,6 +119,10 @@
 
+    option '--strict-semver' do
+      summary _('Whether version ranges should exclude pre-release versions')
+    end
```
AFAICT, this option isn't used in the `install` application. Can it be dropped?

thallgren replied:
No, the setting is picked up by `Puppet::ModuleTool::Applications::Installer` [here](/puppetlabs/puppet/pull/5883/files#diff-90a94f556f7339081d9d810d564e9800R28) and then used [here](/puppetlabs/puppet/pull/5883/files#diff-90a94f556f7339081d9d810d564e9800R63), [here](/puppetlabs/puppet/pull/5883/files#diff-90a94f556f7339081d9d810d564e9800R107), [here](/puppetlabs/puppet/pull/5883/files#diff-90a94f556f7339081d9d810d564e9800R115), [here](/puppetlabs/puppet/pull/5883/files#diff-90a94f556f7339081d9d810d564e9800R194), and [here](/puppetlabs/puppet/pull/5883/files#diff-90a94f556f7339081d9d810d564e9800R209).

joshcooper commented on lib/puppet/face/module/list.rb:
```diff
@@ -17,6 +17,10 @@
+    option '--strict-semver' do
+      summary _('Whether version ranges should exclude pre-release versions')
+    end
+
```
Might be clearer to set the default, like we do for the `puppet generate type --force` flag:

```ruby
default_to { false }
```

thallgren replied:
While I agree that setting defaults closer to a parameter declaration is a good thing, all module sub-command defaults are currently setting defaults in `Puppet::ModuleTool.set_option_defaults options`. I continued to use that pattern here. Changing that pattern deserves a ticket of its own I think.

joshcooper commented on lib/puppet/vendor/pathspec/lib/pathspec/spec.rb:
```diff
@@ -1,4 +1,4 @@
-class Spec
+class BSpec
```
Why did we need to rename this?

thallgren replied:
That shouldn't have made it into the PR. The reason for the rename is that the module_tool tests doesn't run when a class is present that is named `Spec` because the tests introduce a module with that name. I pushed commit where that rename is reverted.
